cns/fsnotify/fsnotify.go

```diff
@@ -0,0 +1,154 @@
+package fsnotify
+
+import (
+	"context"
+	"io/fs"
+	"os"
+
+	"github.com/Azure/azure-container-networking/azure-ipam/logger"
+	"github.com/Azure/azure-container-networking/cns"
+	"github.com/Azure/azure-container-networking/cns/client"
+	"github.com/fsnotify/fsnotify"
+	"github.com/pkg/errors"
+	"go.uber.org/zap"
+)
+
+const (
+	directory = "/deleteIDs"
+)
+
+type Watcher struct {
+	CnsClient *client.Client
+	logger    *zap.Logger
+}
+
+func WatchFs(w *Watcher, path string) error {
+	loggerCfg := &logger.Config{
+		Level:       "debug",
+		Filepath:    "/var/log/fsnotify-watcher.log",
+		MaxSizeInMB: 5, // MegaBytes
+		MaxBackups:  8,
+	}
+	// Create logger
+	fsnotifyLogger, cleanup, err := logger.New(loggerCfg)
+	if err != nil {
+		return errors.Wrapf(err, "failed to setup fsnotify logging")
+	}
+	fsnotifyLogger.Debug("logger construction succeeded")
+	w.logger = fsnotifyLogger
+	defer cleanup()
+
+	// Create new fs watcher.
+	watcher, err := fsnotify.NewWatcher()
+	if err != nil {
+		w.logger.Error("Error creating watcher: ", zap.Error(err))
+	}
+	defer watcher.Close()
+
+	// Start listening for events.
+	go func() {
+		for {
+			select {
+			case event, ok := <-watcher.Events:
+				if !ok {
+					return
+				}
+				w.logger.Info("Watcher: ", zap.Any("event: ", event.Name))
+				if event.Has(fsnotify.Create) {
+					w.logger.Info("created file: ", zap.Any("event: ", event.Name))
+					w.releaseIP(event.Name, path)
+				}
+				if event.Has(fsnotify.Remove) {
+					w.logger.Info("removed file: ", zap.Any("event: ", event.Name))
+				}
+			case err, ok := <-watcher.Errors:
+				if !ok {
+					return
+				}
+				w.logger.Error("Watcher Error: ", zap.Error(err))
+			}
+		}
+	}()
+
+	// Add directory where intended deletes are kept
+	dirpath := path + "/" + directory
+	err = os.MkdirAll(dirpath, 0o755)
+	if err != nil {
+		w.logger.Error("Error making directory: ", zap.Error(err))
```

Comment:
shouldn't this be created in `WatcherAddFile` call? what if cni starts and call del before cns starts?

Reply:
The directory needs to be created in WatchFS before we add it to the watcher to monitor for events. If CNS is deleted the directory with intended deletes is still present and holds the containerIDs that need to be removed next